docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst

```diff
@@ -0,0 +1,1946 @@
+restricts the kinds of changes that a new package version can introduce
+over its prior version.
+
+For example, the simplest kind of data transformation that SCU supports
+is adding a field to a template.
+
+Given the following first version of a template:
+
+.. code:: daml
+
+  template IOU
+    with
+      issuer: Party
+      owner: Party
+      value: Int
+    where
+      signatory issuer
+      observer owner
+
+We can add a new field for currency:
+
+.. code:: daml
+
+  template IOU
+    with
+      issuer: Party
+      owner: Party
+      value: Int
+      -- New field:
+      currency: Optional String
+    where
+      signatory issuer
+      observer owner
+
+With SCU, any new template fields must be optional - templates from the
+old version are automatically upgraded to new versions by setting the
+new field to ``None``. This optional field requirement extends to all
+records in your package. Conversely, newer contracts with this field set
+to ``None`` can be automatically downgraded to previous versions of the
+template in workflows that have not yet been updated.
+
+Automatic Data Upgrades
+~~~~~~~~~~~~~~~~~~~~~~~
+
+When extending data in a daml model, SCU requires the old model be
+representable in the new model. I.e., for extending a record, we can
+only add nullable (``Optional``) fields, so that old data can be represented
+by setting these fields to ``None``. This mentality is continued for
+Variants and Enums, which only allow adding constructors, with some
+other restrictions covered in `Continuing to Write Your Upgrades <#continuing-to-write-your-upgrades>`__. This
+approach is inspired by
+`Protobuf <https://protobuf.dev/programming-guides/proto3/#updating>`__
+and Typescript's ability to ignore `excess
+fields <https://www.typescriptlang.org/docs/handbook/2/objects.html#excess-property-checks>`__
+via ``as``.
+
+Following is a list of places where automatic data upgrades occur:
+
+**Submissions to the ledger-api**
+
+When you submit a command, and do not explicitly specify a package-id to
+use, canton will automatically upgrade the payloads you give to the most
+recent version of the package that is uploaded on the participant. It
+will also use the most recent implementation of any choices you exercise
+directly through the ledger-api. This behavior can be influenced by
+package preference, which will be covered later.
+
+**Fetches/Creates**
+
+Within choices, the package-ids of the payloads we expect for fetches
+and creates are given by the version of the package we depend on. SCU
+will upgrade (or downgrade) the payloads given/stored in the ledger to
+match these versions when used in a choice body.
+
+**Exercises**
+
+Similarly to fetches, the version of the choice implementation that is
+called from a choice body is determined by the package that the choice’s
+package depends on. SCU will automatically upgrade/downgrade the payload
+stored in the ledger to match that of the choice it is calling.
```

Comment:
Missing: result types

Reply:
Note that the engine does not upgrade/downgrade the result types, it will normally be the clients that do the conversion (same logic as the engine).

Reply:
I've reworded this section to forward mention that result types are handled by consuming clients, and to explicitly mention it in consuming clients.